src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -392,61 +674,220 @@ void ShenandoahFreeSet::clear() {
 }
 
 void ShenandoahFreeSet::clear_internal() {
-  _mutator_free_bitmap.clear();
-  _collector_free_bitmap.clear();
-  _mutator_leftmost = _max;
-  _mutator_rightmost = 0;
-  _collector_leftmost = _max;
-  _collector_rightmost = 0;
-  _capacity = 0;
-  _used = 0;
+  _partitions.make_all_regions_unavailable();
 }
 
-void ShenandoahFreeSet::rebuild() {
-  shenandoah_assert_heaplocked();
-  clear();
-
+// This function places all regions that have allocation capacity into the mutator_partition.  Subsequently, we will
+// move some of the mutator regions into the collector partition with the intent of packing collector memory into the
+//  highest (rightmost) addresses of the heap, with mutator memory consuming the lowest addresses of the heap.
+void ShenandoahFreeSet::find_regions_with_alloc_capacity(size_t &cset_regions) {
```

Comment:
There is a public API for prepare_to_rebuild() followed by finish_rebuild().  This public API is exercised by GenShen, which adjusts the sizes of old-gen and young-gen between the two calls.  Single-gen shenandoah does not distinguish between these two steps, because it has no notion of adjusting generation sizes.  Single-gen shenandoah invokes the public api rebuild(), which simply delegates to these two functions.